source/common/http/conn_manager_impl.cc

```diff
@@ -2224,8 +2226,17 @@ Upstream::ClusterInfoConstSharedPtr ConnectionManagerImpl::ActiveStreamFilterBas
 }
 
 Router::RouteConstSharedPtr ConnectionManagerImpl::ActiveStreamFilterBase::route() {
+  return route(nullptr);
+}
+
+Router::RouteConstSharedPtr
+ConnectionManagerImpl::ActiveStreamFilterBase::route(const Router::RouteCallback& cb) {
+  if (cb && parent_.cached_route_.has_value()) {
+    clearRouteCache();
+  }
```

Comment:
Is this actually the behavior you want? It seems a bit different than what `route()` does. I think you would want to just have a callback user call the refresh function if they want to clear it? IMO we should be consistent here.

Reply:
I thought about this for a while, if I returned the cached route to the callback and the callback wanted to override then I'll have to thread this route all the way so it is skipped over or we skip this over in here, otherwise we will end up delivering the same route twice.

Skipping the route over here probably more clean than threading through even cleaner is just clear cache if cb is provided. Let me know what you think.

Reply:
Sorry why can't the filter just call `clearRouteCache` and then call route with a callback?

Reply:
If the caller didn't invoke `clearRouteCache` before invoking route and override the first rout(cached route) we would end up returning it again, we are expecting the caller to so perform additional steps for cleaner behavior, which we could have done easily. However if you feel that both route and and route with cb should have similar implementation in terms of `clearRouteCache` I don't have issues with that will send a fix.

Reply:
> However if you feel that both route and and route with cb should have similar implementation in terms of clearRouteCache I don't have issues with that, I will send in the fix.

Yes let's do that, thanks.

Reply:
One more subtle question, caller invokes route with cb but has not cleared the cache we in

Reply:
IMO if the caller passes a callback to this function and there is a cached route the callback should be ignored. This is consistent with the other variant of the function.

Reply:
Sounds good to me will do so and update the documetation